storage: TestAdminRelocateRange failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/09d51e9f6265ed70caf49385be905606ebf722c7

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestAdminRelocateRange PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1515520&tab=buildLog

```
5      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.800330 199716 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.801034 199716 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
I191001 07:49:09.800434 199715 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] closedts-subscription
I191001 07:49:09.802486 199815 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.803246 199815 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
I191001 07:49:09.804106 199815 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
I191001 07:49:09.806409 199814 util/stop/stopper.go:542  quiescing; tasks left:
4      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.806870 199814 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.811755 199814 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191001 07:49:09.812046 199814 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
I191001 07:49:09.812368 199814 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] closedts-subscription
I191001 07:49:09.838415 199727 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n1: failed to connect to n1 at 127.0.0.1:42423: context canceled
I191001 07:49:09.839823 199726 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n3: failed to connect to n3 at 127.0.0.1:43107: context canceled
I191001 07:49:09.841249 199725 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n2: failed to connect to n2 at 127.0.0.1:33803: context canceled
I191001 07:49:09.844048 199729 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n6: failed to connect to n6 at 127.0.0.1:36383: context canceled
I191001 07:49:09.844501 199728 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n5: failed to connect to n5 at 127.0.0.1:35035: context canceled
I191001 07:49:09.904825 199774 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n3: failed to connect to n3 at 127.0.0.1:43107: context canceled
I191001 07:49:09.905057 199777 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n4: failed to connect to n4 at 127.0.0.1:45871: context canceled
E191001 07:49:09.906461 199779 vendor/google.golang.org/grpc/pickfirst.go:61  pickfirstBalancer: failed to NewSubConn: rpc error: code = Canceled desc = grpc: the client connection is closing
I191001 07:49:09.906014 199776 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n1: failed to connect to n1 at 127.0.0.1:42423: context canceled
I191001 07:49:09.906799 199775 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n2: failed to connect to n2 at 127.0.0.1:33803: context canceled
I191001 07:49:09.906867 199778 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n5: failed to connect to n5 at 127.0.0.1:35035: context canceled
--- FAIL: TestAdminRelocateRange (8.51s)
    require.go:794: 
        	Error Trace:	client_relocate_range_test.go:35
        	            				client_relocate_range_test.go:189
        	            				client_relocate_range_test.go:100
        	            				client_relocate_range_test.go:188
        	Error:      	Received unexpected error:
        	            	while carrying out changes [{ADD_REPLICA n3,s3} {REMOVE_REPLICA n1,s1}]: snapshot failed: (n3,s3):12LEARNER: remote failed to apply snapshot for reason failed to apply snapshot: [n3,s3]: canApplySnapshotLocked: cannot add placeholder, have an existing placeholder range=25 [/Table/Max-/Max) (placeholder) (n4,s4):9
        	Test:       	TestAdminRelocateRange

```

SHA: https://github.com/cockroachdb/cockroach/commits/4dcfd7d8899d90dd1816153bf59b5df647c9bbd3

Parameters:
```
TAGS=
GOFLAGS=-race -parallel=2
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestAdminRelocateRange PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1517125&tab=buildLog

```
W191002 12:09:02.985747 190191 storage/raft_transport.go:634  [n3] while processing outgoing Raft queue to node 5: rpc error: code = Canceled desc = grpc: the client connection is closing:
I191002 12:09:02.986656 190724 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n6: failed to connect to n6 at 127.0.0.1:36375: node unavailable; try another peer
W191002 12:09:02.989048 189848 storage/raft_transport.go:634  [n3] while processing outgoing Raft queue to node 4: rpc error: code = Canceled desc = grpc: the client connection is closing:
I191002 12:09:02.991813 190908 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] closedts-subscription
I191002 12:09:02.992063 190933 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n3: failed to connect to n3 at 127.0.0.1:36457: node unavailable; try another peer
W191002 12:09:02.984372 190815 storage/raft_transport.go:634  [n6] while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W191002 12:09:02.995260 190201 storage/raft_transport.go:634  [n3] while processing outgoing Raft queue to node 5: rpc error: code = Canceled desc = grpc: the client connection is closing:
I191002 12:09:02.994120 190934 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n2: failed to connect to n2 at 127.0.0.1:32773: context canceled
I191002 12:09:02.996354 190906 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I191002 12:09:02.996678 190911 util/stop/stopper.go:542  quiescing; tasks left:
5      [async] closedts-subscription
W191002 12:09:02.997244 190092 storage/raft_transport.go:634  [n5] while processing outgoing Raft queue to node 6: rpc error: code = Canceled desc = grpc: the client connection is closing:
I191002 12:09:02.997445 190906 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
I191002 12:09:02.997609 190911 util/stop/stopper.go:542  quiescing; tasks left:
3      [async] closedts-subscription
I191002 12:09:02.997961 190910 util/stop/stopper.go:542  quiescing; tasks left:
4      [async] closedts-subscription
I191002 12:09:02.998490 190911 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] closedts-subscription
I191002 12:09:02.998685 190910 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] closedts-subscription
I191002 12:09:02.999102 190910 util/stop/stopper.go:542  quiescing; tasks left:
1      [async] closedts-subscription
I191002 12:09:03.045851 190820 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n4: failed to connect to n4 at 127.0.0.1:40573: context canceled
I191002 12:09:03.046425 190819 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n3: failed to connect to n3 at 127.0.0.1:36457: context canceled
I191002 12:09:03.046662 190950 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n1: failed to connect to n1 at 127.0.0.1:45227: context canceled
I191002 12:09:03.046553 190818 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n5: failed to connect to n5 at 127.0.0.1:40483: context canceled
I191002 12:09:03.049659 190949 rpc/nodedialer/nodedialer.go:160  [ct-client] unable to connect to n2: failed to connect to n2 at 127.0.0.1:32773: context canceled
--- FAIL: TestAdminRelocateRange (8.67s)
    require.go:794: 
        	Error Trace:	client_relocate_range_test.go:35
        	            				client_relocate_range_test.go:189
        	            				client_relocate_range_test.go:100
        	            				client_relocate_range_test.go:188
        	Error:      	Received unexpected error:
        	            	while carrying out changes [{ADD_REPLICA n5,s5} {REMOVE_REPLICA n4,s4}]: snapshot failed: (n5,s5):13LEARNER: remote couldn't accept LEARNER snapshot c8325d24 at applied index 102 with error: [n5,s5],r25: cannot apply snapshot: [n5,s5]: canApplySnapshotLocked: cannot add placeholder, have an existing placeholder range=25 [/Table/Max-/Max) (placeholder) (n3,s3):12
        	Test:       	TestAdminRelocateRange

```

--- a/pkg/storage/client_relocate_range_test.go
+++ b/pkg/storage/client_relocate_range_test.go
@@ -31,12 +31,20 @@ func relocateAndCheck(
 	tc *testcluster.TestCluster,
 	startKey roachpb.RKey,
 	targets []roachpb.ReplicationTarget,
-) {
-	require.NoError(t, tc.Servers[0].DB().AdminRelocateRange(context.Background(), startKey.AsRawKey(), targets))
+) (retries int) {
+	testutils.SucceedsSoon(t, func() error {
+		err := tc.Servers[0].DB().
+			AdminRelocateRange(context.Background(), startKey.AsRawKey(), targets)
+		if err != nil {
+			retries++
+		}
+		return err
+	})
 	desc, err := tc.Servers[0].LookupRange(startKey.AsRawKey())
 	require.NoError(t, err)
 	requireDescMembers(t, desc, targets)
 	requireLeaseAt(t, tc, desc, targets[0])
+	return retries
 }
 
 func requireDescMembers(
@@ -94,10 +102,10 @@ func TestAdminRelocateRange(t *testing.T) {
 	}
 	var intercepted []intercept
 
-	requireNumAtomic := func(expAtomic int, expSingle int, f func()) {
+	requireNumAtomic := func(expAtomic int, expSingle int, f func() (retries int)) {
 		t.Helper()
 		intercepted = nil
-		f()
+		retries := f()
 		var actAtomic, actSingle int
 		for _, ic := range intercepted {
 			if ic.err != nil {
@@ -109,6 +117,7 @@ func TestAdminRelocateRange(t *testing.T) {
 				actSingle++
 			}
 		}
+		actAtomic -= retries
 		require.Equal(t, expAtomic, actAtomic, "wrong number of atomic changes: %+v", intercepted)
 		require.Equal(t, expSingle, actSingle, "wrong number of single changes: %+v", intercepted)
 	}
@@ -137,8 +146,8 @@ func TestAdminRelocateRange(t *testing.T) {
 	{
 		targets := tc.Targets(1, 0, 2)
 		// Expect two single additions, and that's it.
-		requireNumAtomic(0, 2, func() {
-			relocateAndCheck(t, tc, k, targets)
+		requireNumAtomic(0, 2, func() int {
+			return relocateAndCheck(t, tc, k, targets)
 		})
 	}
 
@@ -151,16 +160,16 @@ func TestAdminRelocateRange(t *testing.T) {
 		// Should carry out three swaps. Note that the leaseholder gets removed
 		// in the process (i.e. internally the lease must've been moved around
 		// to achieve that).
-		requireNumAtomic(3, 0, func() {
-			relocateAndCheck(t, tc, k, targets)
+		requireNumAtomic(3, 0, func() int {
+			return relocateAndCheck(t, tc, k, targets)
 		})
 	}
 
 	// s4 (LH) s5 s6 ---> s5 (LH)
 	// Pure downreplication.
 	{
-		requireNumAtomic(0, 2, func() {
-			relocateAndCheck(t, tc, k, tc.Targets(4))
+		requireNumAtomic(0, 2, func() int {
+			return relocateAndCheck(t, tc, k, tc.Targets(4))
 		})
 	}
 
@@ -168,8 +177,8 @@ func TestAdminRelocateRange(t *testing.T) {
 	// Lateral movement while at replication factor one. In this case atomic
 	// replication changes cannot be used; we add-then-remove instead.
 	{
-		requireNumAtomic(0, 2, func() {
-			relocateAndCheck(t, tc, k, tc.Targets(2))
+		requireNumAtomic(0, 2, func() int {
+			return relocateAndCheck(t, tc, k, tc.Targets(2))
 		})
 	}
 
@@ -177,16 +186,16 @@ func TestAdminRelocateRange(t *testing.T) {
 	// A grab bag.
 	{
 		// s3 -(add)-> s3 s2 -(swap)-> s4 s2 -(add)-> s4 s2 s1 (=s2 s4 s1)
-		requireNumAtomic(1, 2, func() {
-			relocateAndCheck(t, tc, k, tc.Targets(1, 3, 0))
+		requireNumAtomic(1, 2, func() int {
+			return relocateAndCheck(t, tc, k, tc.Targets(1, 3, 0))
 		})
 		// s2 s4 s1 -(add)-> s2 s4 s1 s6 (=s4 s2 s6 s1)
-		requireNumAtomic(0, 1, func() {
-			relocateAndCheck(t, tc, k, tc.Targets(3, 1, 5, 0))
+		requireNumAtomic(0, 1, func() int {
+			return relocateAndCheck(t, tc, k, tc.Targets(3, 1, 5, 0))
 		})
 		// s4 s2 s6 s1 -(swap)-> s3 s2 s6 s1 -(swap)-> s3 s5 s6 s1 -(del)-> s3 s5 s6 -(del)-> s3 s5
-		requireNumAtomic(2, 2, func() {
-			relocateAndCheck(t, tc, k, tc.Targets(2, 4))
+		requireNumAtomic(2, 2, func() int {
+			return relocateAndCheck(t, tc, k, tc.Targets(2, 4))
 		})
 	}
 }

